### ISSUE TITLE
Reconsider useObserver

### ISSUE THREAD

**Issue by davezuko** [mobxjs/mobx-react-lite#323]
Hey all. Congratulations on the recent releases! I've been happily chugging along with mobx for a while now, and after a relatively pleasant migration to mobx@6 I was a bit disappointed to see that `useObserver` got deprecated in this package. I read through https://github.com/mobxjs/mobx-react-lite/pull/306 where this was discussed (much too late, unfortunately). In the interest of soliciting opinions from others in the community, I thought I'd raise it as a standalone issue.

While I recognize the potential for abuse and user-error, I think `useObserver`'s overall cost/benefit ratio warrants its inclusion. To draw a parallel, mobx core thoroughly documents common anti-patterns, and I personally think that that is entirely sufficient. The same applies here: there are real benefits to `useObserver` when compared to the alternatives, and I think that with sufficient documentation it would be reasonably safe to include.

The benefits of `useObserver` over `observer(MyComponent)`

1. No unnecessary component wrapping, which means: no need to hoist statics, handle ref forwarding, etc. That's less _stuff_ for mobx to handle, and a slightly smaller runtime footprint. The community ts this point has largely moved on past the HoC pattern, and I think we're better for it. Relatedly:
2. It's composable. I can define a utility hook for a root store, for example. See:

```js
const useRootStore = (render) => {
  return useObserver(() => {
    const store = useContext(MyStoreContext)
    return render(store)
  })
}
```

3. More ergonomic exports. With the HoC style, to get proper displayNames you need to either:
    a. Switch from an arrow function to a regular named function to add a `displayName`
    b. Define your component first, then later reassign it (e.g. `MyComponent = observer(MyComponent)`) 
    c. Set `displayName` yourself (with proper development flags)
4. A less-cluttered React tree, which is (at least in my books) a developer/debugging convenience.

As far as `<Observer>` goes, it's quite clunky ergonomically, and stands out more now that "render props" have fallen out of fashion. More subjectively, I'm quite averse to using React components as _utilities_; I want the React tree to reflect the final UI as much as possible in practice. Hooks really help with this.

## Compromise?

If the goal is to encourage people to _exclusively_ use it for rendering, and not data selection, perhaps consider naming it `useObservedRender` or asserting that it returns a `ReactNode` in development. Or, go so far as to encourage the use of the `obsever(Component) and <Observer />` API's, but also retain the `useObserver` API for those who wish to use it.

In conclusion, I readily admit that there's a lot about mobx that I'm not an expert on. Nonetheless, `useObserver` is, in my opinion, one of the most straightforward implementations of a React/Mobx integration that I can think of, and I think it should remain in core. If all else fails, I know that I'll at least be re-implementing it in userland for usage in our projects. Thanks again!

**Comment by mweststrate** [comment]
I don't think any new arguments are brought to the table here. That it looks nicer is a pretty poor argument imho to introduce the need to all consumers to understand the difference. Beyond that your above example reeks like an anti-pattern, hooks should be called unconditionally, which is not guaranteed by your nesting without knowing the semantics of useObserver. I can actually imagine the standard React lint rules will fire on your example, although I didn't try that. 

I think for me your request can be kind of paraphrased as: "A kitchen knife can be used to chop strawberries. But can we use swords as well? Yes it is much more dangerous, but then again, it looks super cool!" (I don't mean that condescending, but witty / light hearted ;-)). 

> To draw a parallel, mobx core thoroughly documents common anti-patterns, and I personally think that that is entirely sufficient. 

That would be my dream. In practice people barely read docs, and they often file issues before reading docs in depth. This has thought mew over the years to be really wary to put too powerful tools in hands of package consumers, because it usually fires back in the form of a high maintenance burden. Trust me, I'd wish "use this at your own risk" warnings would actually work 😅

FWIW if you insist on having the utility, I wouldn't roll your own, but rather apply patch-package to force an export, so that your logic doesn't get out of date with the original implementation :) (or simply wrap Observer, that will probably work in most cases, but possibly not in all)

**Comment by davezuko** [comment]
Thanks for the quick reply! I'll admit I was hoping for a different response, but I hope you will at least leave the issue open for another few days to see if anybody else agrees (or disagrees :D).

> That it looks nicer is a pretty poor argument imho 

I considered omitting the "looks nicer" parts of the argument, but included them for completeness. I regret that now, and so I'll reiterate that I believe the real benefit of `useObserver` to be its composability, full-stop. I hope that maybe this point can be addressed further. Which brings up:

> Beyond that your above example reeks like an anti-pattern, hooks should be called unconditionally, which is not guaranteed by your nesting without knowing the semantics of useObserver.

This caught me off guard, since I don't follow how the example is an anti-pattern. If `useObserver` follows the rules of hooks (which it does), then composing it inside of another hook is a non-issue. There are no conditionals (short of anyone breaking the rules of hooks inside of "render", which would be an issue anyways). I'm happy to be enlightened if I'm mistaken, but this seems to me to be precisely the thing hooks were designed for.

> I think for me your request can be kind of paraphrased as: "A kitchen knife can be used to chop strawberries. But can we use swords as well? Yes it is much more dangerous, but then again, it looks super cool!" 

No offense taken. However, what I meant to get across that mobx itself offers (in my opinion) a fair number of more "sword-like" public API's that trust that the user knows what he/she's doing. I meant to draw a parallel here.

> That would be my dream. In practice people barely read docs, and they often file issues before reading docs in depth

Totally understand. I happen to be a person who _does_ read through all of the docs, but I can see how frustrating this can be as a maintainer, and I don't mean to imply that you should be forced to support problematic features. I was hoping to point out that there's a grey area here, where for the library to be practical to "power" users there has to be some element of trust.

> FWIW if you insist on having the utility, I wouldn't roll your own, but rather apply patch-package to force an export, so that your logic doesn't get out of date with the original implementation :) (or simply wrap Observer, that will probably work in most cases, but possibly not in all) 

Thanks, I'll consider that :). And, as always, I truly appreciate the work you and others have done on mobx.

**Comment by FredyC** [comment]
I was probably the biggest "defender" of `useObserver` to keep in public, but I no longer think it's needed. I mean sure, HOC can be annoying for a few reasons you have mentioned (especially exports), but otherwise, it's a clear API without any surprises. The `useObserver` does need a deeper understanding of what it does and how it can be used.

> This caught me off guard, since I don't follow how the example is an anti-pattern. If `useObserver` follows the rules of hooks (which it does), then composing it inside of another hook is a non-issue.

It's actually the number one rule: **Call Hooks from React function components.**. In your case, you are calling hook from another hook and that's against that rule. Let's put aside it will work simply because it's synchronous flow.

However, I am thinking we could actually keep it exported (without documenting it), and additionally, when used for the first time, it would throw a hard error unless you call something like `enableUseObserverWithAllRisks` before that. The goal is to make it concise decision and clearly stating we won't be supporting the use of that. And if course when used internally by `observer/Observer`, it won't require anything like that. It would be a little dark area of the lib for power users :)

If you are ok with something like that and you are able to work on a PR, I would give that a gree light. Otherwise, I don't think anyone else would be willing to work on that :)

**Comment by vkrol** [comment]
> It's actually the number one rule: Call Hooks from React function components.. In your case, you are calling hook from another hook and that's *against that rule*.

Why? Calling a hook from another hook is absolutely correct code.
> Only call Hooks from React function components. Don’t call Hooks from regular JavaScript functions. (*There is just one other valid place to call Hooks — your own custom Hooks*. We’ll learn about them in a moment.)

**Comment by mweststrate** [comment]
I think @FredyC is referring to https://reactjs.org/docs/hooks-rules.html#only-call-hooks-at-the-top-level:

> Don’t call Hooks inside loops, conditions, or nested functions.

The hook call here is not a top level statement, but yes it does technically work correct, but only if the assumption is made that useObserver calls it fn always, immediately, and unconditionally. That is the case currently, but things like suspense, debounced rendering, microtasking, worker rendering etc would all kill that assumption. We don't guarantee anywhere that tracking in useObserver happens immediately. So it is not unlikely that in a patch or minor version this changes because we apply some optimization; thereby breaking this kind of utilities. In contrast, the following achieves exactly the same as far as I can see, has none of the downsides, and is imho actually clearer to read:

```
const useRootStore = (render) => {
  const store = useContext(MyStoreContext)
  return <Observer>{() => render(store)}</Observer>
}
```

**Comment by urugator** [comment]
> the primary benefit of useObserver to be its composability, full-stop

The composability of `useObserver` is at least questionable.
Hooks must not contain dynamic logic (loops/conds).
Rendering on the other hand is generally dynamic - so logically you cannot have a rendering inside a hook (unless you make some assumptions about the rendering logic).
So given that `useObservable` is intended for rendering, it's not really composable.
Generally speaking hooks are not intended for rendering (exactly because they wouldn't compose), the components are.
Actually I think that `useObservable` shouldn't be considered a hook. It's a wrong abstraction for the problem it solves no matter how attractive it looks. A "proper" version could look like this:
```javascript
function cmp() {
  // a hook
  const render = useObserver();
  
  // not a hook
  return render(() => {

  })
}
```
Which is less attractive and raises other questions/gotchas/issues.

Another problem with this "composability" is that you loose the benefits of mobx's fine grained subscriptions - we cannot invalidate/rerender only single `useObserver` we always invalidate the whole component.  
It's not neccessarily always wrong, because in the end it's just an optimization, but it somehow goes against the mobx design, so probably not something we want to encourage.

But really the biggest problem is how easy it is for a user to introduce an error, without any hints for solving it. With `observer` all that can happen is that you forget to apply it.
Also it may lead to incorrect conclusions about how mobx works or how to use it efficiently.

> A less-cluttered React tree, which is (at least in my books) a developer/debugging convenience.

I think you can only speak about tree cluttering if nodes in the tree are non-visual and provides just logic/utility.
Afaik we don't add any additional nodes to the tree. Except memo that you would have to (or should) add anyway.

**Comment by davezuko** [comment]
Fair points made by all. I appreciate all of the feedback and rest my case for now.

I still have some disagreements about the whole "it violates rules of hooks" arguments; however, given the other reasonable points made by everyone above I don't intend to die on this hill. With that, I'll close this for now. If anybody else feels more strongly than me please feel free to bump the issue.

---

Just for the sake of discussion and context, I did want to reply to the following:

> Rendering on the other hand is generally dynamic - so logically you cannot have a rendering inside a hook

I stared at this for a long time and still couldn't quite come to grips with it. I'm intimately familiar with the rules of hooks: React needs the same hooks to be called in the same order each render. Beyond that, your component is free to be as dynamic as it pleases. So, to me it followed that a render function that follows the rules of hooks should be able to be wrapped in another hook.

I still don't see any issue with any of that. What I did realize, though, is perhaps what @mweststrate alluded to earlier, and it has nothing to do with the dynamicism of a render function (because it _must_ follow the rules of hooks regardless). Because `useObserver` takes a render function as a callback, it's possible that it can invoke it (and thus any hooks) at any point (i.e. outside of the React lifecycle, which is a big no-no). In this way, we are relying on `useObserver`'s implementation details, trusting that it won't call `render` in a way that breaks the rules of hooks. I considered this implementation detail necessary to the very functioning of the hook, and therefore not liable for change, but it seems this isn't something we could/should count on?

---

> React devtools have a filtering capability for a while, they even allow hiding all HOC. So that's an argument of the past ;)

I... somehow never knew this, and now feel like a dummy. Thank you, @FredyC.

**Comment by urugator** [comment]
My reasoning goes like this:
The **only concern** of the `useObservable` hook is **rendering** - it wraps a rendering logic.
So when we talk about **composing `useObservable`**, we can only talk about **composing rendering logic**.
The **rendering** logic is supposed to **contain conditions and loops** (that's why we use react after all).
But the **hooks cannot contain conditions and loops**.
So either it **does not** support rendering very well, because we cannot use conditions and loops in the redering, which is sad, because the rendering is only thing it does.
Or it **does** support any kind of rendering logic, but then it's non-composable (with itself or other hooks), because it contains conditions and loops.

I am only questioning the composability, considering the purpose of the hook and restrictions of hooks.
In other words, if we know the rules of hooks, does it make sense to place a rendering logic in them? Is hook a good place for rendering logic, if we know it cannot contain conditions and loops...? I don't claim that doing so automatically violates something, but it has some implications in regards to composability.

**Comment by JabX** [comment]
I know this issue has been closed for 2 weeks now, but I'd like to add a few more things to the discussion, as I am very much in favor of keeping `useObserver` exported somewhere too, and I actually got convinced by @FredyC a while ago.

My main issue with encouraging the use of a `observer` HoC is that people will always use anonymous functions do describe their components, like it is shown in the docs. Not having a proper name for the component is a problem for the dev tools, stack traces, and so on. No one will ever bother writing something like `export const MyComponent = observer(function MyComponent(props)) { ... })`, so you'll always have to explain people why they need to do it anyway (and no, default exports are never the solution).

My other issue with the `observer` HoC is that, in Typescript, inferring the type of a component wrapped in a HoC doesn't work, at least not consistently in my experience. For instance, I had this problem all the time with generic list components that have a "LineComponent" prop whose type depends on the type of the list. While I'd agree that defining components on the fly during render isn't the greatest idea, most of the time it's perfectly fine, especially since that component doesn't have to rerender again when the list changes.

Using `useObserver` solves both problems (since there's no HoC) and it hasn't been too hard for my team to adopt it. In my opinion, its rules are simple : the JSX you return has to be wrapped by `useObserver`, and any observable you read during render must be read inside `useObserver`. It does a lot less than `observer` and makes people understand what they're doing.

I agree that it's not the easiest way to get going and that it can get confusing for someone who didn't read the docs. I also agree that its nature as a hook is weird. But at the very least its a "low level" utility that's an export away from being available (I know it's still available behind a deprecation message, but that's bound to change), so it doesn't cost much to keep it. It can be hidden behind a big-ass warning like `observe` and `intercept` (which are still very useful, I remember reading that they were considered for deprecation once btw?) for instance. I mean, it still looks a lot better to me than having to resort to patching the library to add the export or, worse, to reimplement it in user-land with `<Observer>`.
